plugin.video.dr.dk.bonanza/resources/lib/addon.py

```diff
@@ -0,0 +1,228 @@
+#      Copyright (C) 2014 Tommy Winther, termehansen
+
+#
+#  https://github.com/xbmc-danish-addons/plugin.video.drnu#
+#
+#  This Program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 2, or (at your option)
+#  any later version.
+#
+#  This Program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+#  GNU General Public License for more details.
+#
+#  You should have received a copy of the GNU General Public License
+#  along with this Program; see the file LICENSE.txt.  If not, write to
+#  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+#  http://www.gnu.org/copyleft/gpl.html
+#
+import os
+import re
+import sys
+import requests
+import requests_cache
+from html import unescape
+import buggalo
+
+import xbmc
+import xbmcgui
+import xbmcplugin
+import xbmcaddon
+import json
+
+try:
+    # python 2
+    from urlparse import parse_qsl
+except:
+    from urllib.parse import parse_qsl
+
+BASE_URL = 'http://www.dr.dk/bonanza/'
+tr = xbmcaddon.Addon().getLocalizedString
+addon_path = xbmcaddon.Addon().getAddonInfo('path')
+addon_name = xbmcaddon.Addon().getAddonInfo('name')
+ICON = os.path.join(addon_path, 'resources', 'icon.png')
+FANART = os.path.join(addon_path, 'resources', 'fanart.jpg')
+CACHE = os.path.join(addon_path, 'requests_cache')
+
+def make_notice(object):
+    xbmc.log(str(object), xbmc.LOGINFO )
+
+class BonanzaException(Exception):
+    pass
+
+
+class Bonanza(object):
+    def __init__(self, plugin_url, plugin_handle):
+        self._plugin_url = plugin_url
+        self._plugin_handle = plugin_handle
+
+        #cache expires after: 86400=1 day   604800=7 days
+        requests_cache.install_cache(CACHE, backend='sqlite', expire_after=86400 )
+        requests_cache.remove_expired_responses()
+
+    def search(self):
+        make_notice('search')
+        keyboard = xbmc.Keyboard('', tr(30001))
+        keyboard.doModal()
+        if keyboard.isConfirmed():
+            html = self._downloadUrl('http://www.dr.dk/bonanza/sog?q=' + keyboard.getText().replace(' ', '+'))
+
+            items = list()
+            pattern = '<a.*?href="/bonanza/(serie/.*?)".*?' \
+                      'data-src="(//asset\.dr\.dk/[^"]+)".*?' \
+                      '<h3.*?>([^<]+)</h3>.*?<p>([^<]+)</p>'
+
+            for m in re.finditer(pattern, html, re.DOTALL):
+                url = BASE_URL + m.group(1)
+                image = 'http:' + m.group(2)
+                title = unescape(m.group(3))
+                description = unescape(m.group(4))
+
+                infoLabels = {
+                    'title': title,
+                    'plot': description,
+                    'studio': addon_name}
+
+                item = xbmcgui.ListItem(infoLabels['title'])
```

Comment:
You might want to add offscreen=True to speed up your add-on. See https://codedocs.xyz/xbmc/xbmc/group__python__xbmcgui__listitem.html#ga15bcb930a4f91dd97f8ed7c633a20618